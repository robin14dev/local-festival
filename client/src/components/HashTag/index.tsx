import React, { useEffect, useRef, useState } from "react";
import { Container, Button } from "./styled";
import { AxiosError } from "axios";

type HashtagProps = {
  onSearch: onSearchFunc;
  query: string | null;
  // setIsTag: React.Dispatch<React.SetStateAction<boolean>>;
};

const tagsArr = [
  {
    text: "봄나들이",
  },
  {
    text: "불빛축제",
  },
  {
    text: "역사탐방",
  },
];

type TagData = { text: string };

const HashTag = ({ onSearch, query }: HashtagProps) => {
  const [tags, setTags] = useState<TagData[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string>("");
  const [curTag, setCurTag] = useState(query);

  // useEffect(() => {
  //   setCurTag(query);
  // }, [query]);

  const fetchTags = async () => {
    function delay(ms: number) {
      const promise = new Promise((resolve, reject) => setTimeout(resolve, ms));
    }
    try {
      setLoading(true);
      await delay(5000);
      setTags(tagsArr);
    } catch (error: unknown) {
      setError("태그를 불러오는데 실패했습니다");
      if (error instanceof AxiosError) {
        console.log(error);
        setError("태그를 불러오는데 실패했습니다");
      }
    } finally {
      setLoading(false);
    }
  };
  useEffect(() => {
    fetchTags();
  }, []);

  return (
    <Container>
      {loading && "축제 태그를 불러오고 있습니다"}
      {!!error && error}
      {tags.map((tag) => (
        <Button
          active={curTag === tag.text}
          onClick={() => {
            onSearch(tag.text);
            setCurTag(tag.text);
            // setIsTag(true);
          }}
          key={tag.text}
        >
          #{tag.text}
        </Button>
      ))}
    </Container>
  );
};

export default HashTag;
