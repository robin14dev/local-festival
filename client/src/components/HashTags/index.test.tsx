import React from "react";
import {  render, screen, waitFor } from "@testing-library/react";
import HashTag from ".";
import userEvent from "@testing-library/user-event";


describe("<HashTag>", () => {

  test("컴포넌트가 성공적으로 렌더링 됩니다.", () => {
    const mockTagData = [{ text: "tag1" }, { text: "tag2" }, { text: "tag3" }];

    const { container } = render(
      <HashTag onSearch={jest.fn()} tagData={mockTagData} />
    );
    const hashTags = screen.getAllByRole("button");
    
    expect(hashTags).toHaveLength(mockTagData.length);
    expect(hashTags[0].textContent).toMatch(/^#+/)
    expect(container).toMatchSnapshot()
  });
  test("해시태그를 클릭시 해시태그의 텍스트가 인자로 들어간 콜백함수가 호출됩니다.", async() => {
    const mockTagData = [{ text: "tag1" }, { text: "tag2" }, { text: "tag3" }];
    const mockFunc = jest.fn()
    const user = userEvent.setup();
   render(
      <HashTag onSearch={mockFunc} tagData={mockTagData} />
    );
    const hashTags = screen.getAllByRole("button");
    for (const hashTag of hashTags) {
      await user.click(hashTag);
      
      expect(mockFunc).toHaveBeenCalledTimes(1);
      expect(mockFunc).toHaveBeenCalledWith(hashTag.textContent?.slice(1))
      mockFunc.mockClear()
    }
  }) 
  test("해시태그에 마우스를 가져다 대면 해당 해시태그의 스타일이 변경됩니다", async () => {
    const mockTagData = [{ text: "tag1" }, { text: "tag2" }, { text: "tag3" }];

    const user = userEvent.setup()
 render(
      <HashTag onSearch={jest.fn()} tagData={mockTagData} />
    );
    const hashTags = screen.getAllByRole("button");
    const hashTag = hashTags[0]
 
    let isHover = false
  
    hashTag.addEventListener('mouseover', () => {
      isHover = true
    })
    hashTag.addEventListener('mouseout', () => {
      isHover = false
    })
    expect(isHover).toBeFalsy()
    waitFor(() => {
      expect(hashTag).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})
    })

    await user.hover(hashTag)
  
    expect(isHover).toBeTruthy()
    expect(hashTag).toHaveStyle({  backgroundColor: '#6268ff', color : 'white' 
    })
  
    await user.unhover(hashTag)
  
    expect(isHover).toBeFalsy()
    waitFor(() => {
      expect(hashTag).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})
    })
  });
  test("해시태그를 클릭시 해당 요소의 스타일이 변경됩니다. 이후, 다른 해시태그를 클릭시, 클릭 되어있던 이전 해시태그의 스타일이 해제됩니다.", async() => {
    const mockTagData = [{ text: "tag1" }, { text: "tag2" }, { text: "tag3" }];
    const mockFunc = jest.fn()
    const user = userEvent.setup();
    render(
      <HashTag onSearch={mockFunc} tagData={mockTagData} />
    );
    const hashTags = screen.getAllByRole("button");
    const hashTag1= hashTags[0]
    const hashTag2 = hashTags[1];
    waitFor(() => {
      expect(hashTag1).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})
      expect(hashTag2).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})
    })
    await user.click(hashTag1)

    waitFor(() => {
      expect(hashTag1).toHaveStyle({  backgroundColor: '#6268ff', color : 'white' 
    })
    expect(hashTag2).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})

    })
    await user.click(hashTag2)
    waitFor(() => {
      expect(hashTag1).toHaveStyle({  backgroundColor: '#f5f6fa', color : 'black'})
      expect(hashTag2).toHaveStyle({  backgroundColor: '#6268ff', color : 'white' 

    })
  })
})
});
