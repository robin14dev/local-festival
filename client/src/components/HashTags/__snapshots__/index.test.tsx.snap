// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HashTag> 컴포넌트가 성공적으로 렌더링 됩니다. 1`] = `
<div>
  <div
    class="sc-bcXHqe ekAJIk"
  >
    <button
      class="sc-gswNZR gbfhpD"
    >
      #
      tag1
    </button>
    <button
      class="sc-gswNZR gbfhpD"
    >
      #
      tag2
    </button>
    <button
      class="sc-gswNZR gbfhpD"
    >
      #
      tag3
    </button>
  </div>
</div>
`;
